New Rust attribute to support embedding debugger visualizers (#3191)
rust-lang/rfcs: text/0000-natvis.md

enabling Rust developers to package debugger visualization scripts with their
crates.

# Motivation
[motivation]: #motivation

Most, if not all, Rust developers will at some point have to debug an issue
in their crate. Trying to view types as they are laid out in memory is not
always the most telling. Furthermore when viewing types from external crates,
the information is even harder to interpret. 

Many languages and debuggers enable developers to control how a type is
displayed in a debugger. These are called "debugger visualizations" or "debugger
views". Debugger views are merely a convenience for some types, such as
`Vec<T>`, but are essential for types such as `HashMap<T>`, where non-trivial
logic is needed in order to correctly display the contents of a type.

Currently, Rust provides visualizations for a handful of types defined in its
standard library via `.natvis` files or python scripts. However, this support
is inflexible; updating it requires modifying the Rust toolchain itself,
and either using a local build of the toolchain or waiting for a new upstream
build of the toolchain. It is not feasible for developers of ordinary crates

> to update the Rust toolchain, solely to add visualizations for their crates.

Reviewer: It would be useful to explain the platform constraints of natvis files here, I think. Perhaps also give an example of what something like HashMap looks like with and without the visualisation applied (put another way, this RFC is assuming that the reader has some knowledge of debugger implementation details, but would be improved if it only assumed the reader has used a debugger)

  then the visualizations of those external crates will "just work" when viewed
  under a debugger without the need of any manual configuration.
* Supports existing debugging visualization systems. We do not propose to
  define a new debugger visualization system; that would be a tremendous
  undertaking, and would ignore the value of existing systems.
* No impact on code quality or size.
* No impact on crates that do not use debugger visualizations.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

This RFC explores making Natvis debugger visualizations extensible, in Rust.
The scenario that we want to enable is:

* Alice publishes a crate, say, `cool_stuff`. Alice wrote debugger
  visualizations for `cool_stuff`, and included them in the crate.
* Bob is writing a new Rust application. Deep in the crate dependency graph of
  Bob's application, some crate uses `cool_stuff`.
  (Bob is not even aware of the existence of debugger visualizations.)
* While Bob is debugging the application, and examining data structures,
  he comes across an instance of `cool_stuff::CoolType` in the debugger.

>   Because Rust and the debugger know about the visualizations that Alice wrote,

Reviewer: I'm not clear on what "Rust" knowing about visualisations means here since there is no Rust runtime or Rust debugger. Presumably, it means that the Rust compiler has built-in some knowledge of the visualisations into the debuginfo in the binary? I think it would be beneficial to be precise here.

Author: I'll clarify this to be precise.

    for captures in rx.captures_iter(input) {
        let my_match: Match = captures.get(1).unwrap();
        do_some_work(my_match.as_str());
    }
}
```

Let's say that Carol is debugging the app, there's a problem in
`do_some_work()`. (Perhaps some code path has triggered a panic.) Carol wants
to look at the state of the app, inside the `find_c_defines` function,
and she specifically wants to see what the state of `captures` is. So she
selects the `find_c_defines` call frame and looks at the local variables
window.

Unfortunately, the debugger's view of the `captures` variable does not give
her any useful information at all. It shows only something like:

```text
> Variables
  > captures: {...}
    > text: "...the entire input text..."
    > locs: {...}
      > __0: (4) vec![None, None, None, None]
      > named_groups: (refs:2) size=0, capacity=1
        > [raw]: alloc::sync::Arc<std::collections::hash::map::HashMap<...>>
          > ptr: {pointer:0xNNNNNNNN}
            > pointer: {...}
              > strong: {...}
              > weak: {...}
              > data: size=0, capacity=1
                > base: {...}
                  > hash_builder: {...}
        ...
```

> The debugger shows the structure of the data, not its meaning. It is useless.

Reviewer: ```suggestion
The debugger shows the structure of the data, not its meaning. It is not very useful for Carol.
```

extract from this variable. In order to meaningfully understand what the `Captures`
type is truly trying to tell us, it would be very helpful to visualize this data
differently in the debugger.

What we _want_ is something like this:

```text
> Variables:
  > captures: {...}
    > $1: "SOME_CONSTANT"
    > $2: "42"
    > $3: "// some developer comment"
```

This RFC will describe how to support adding Natvis visualizations which is supported by:

* The Windows Debugger (WinDbg)
* Visual Studio Debugger.

It should be easy for crate developers to add debugger visualizations to their
crates.

## Supporting Natvis

This section describes how Microsoft's Natvis is supported in Rust.

To use Natvis, developers write XML documents that describe how debugger types
should be displayed. (See: https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019)
The Natvis XML files provide patterns, which match type names, and for matching
types, a description of how to display those types. This allows for some
limited support for generic types.

When writing Natvis files for C++, developers write a standalone XML document

> and add it to their project. The build system (such as Visual Studio) knows

Reviewer: "add it to their project" what does "add" mean here? What is the "project"?

Reviewer 2: The phrasing is fairly standard in the C++ / Visual Studio community.

Reviewer: Sure, but this is not the C++ / Visual Studio community, so it would be good to explain it :-)

Reviewer: To clarify, I'm trying to understand exactly how natvis files are handled by VS - does this mean that VS stores a record of the filename in some project metadata (roughly equivalent to a manifest file in that it would be committed to version control, etc) and that this record is used when building the project? Although this sounds kind of obvious, one could imagine alternatives such as the project being a developer's local view, or "adding" a natvis file meaning that it is copied into source code or compiled into an intermediate format or something.

Author: > "add it to their project" what does "add" mean here? What is the "project"?

I've updated this information and described what it would mean to add Natvis to a project within Visual Studio and refactored the structure of this.

how to package the Natvis file into the debug data (the PDB file) for the
project, and the debugger knows how to find the Natvis XML in each PDB file.

> 

Reviewer: Could you clarify here how the natvis visualisations make their way from the files known to Cargo to a compiler Rust binary?

Author: Yes, I will add more clarification here. Thanks.

Developers can add one or more standalone Natvis XML files to their crate.

> The `Cargo.toml` file specifies the name of these Natvis files. This is the

Reviewer: Is only the name specified? What does the declaration look like? If we only need the name, why not list the contents of a directory?

Author: The path to the .natvis file would be specified. I'll update the language to reflect this. Currently this RFC is aimed at listing files. Presumably in the future this could be updated to include support for directories but just as the auto-discovering of natvis files it's out of scope for this initial RFC.

easiest way to add Natvis support to a project.

> The advantage of a standalone XML document is that this process is already

Reviewer: I'm not clear here. Is the XML intrinsic to natvis, or are there alternative representations?

Reviewer 2: All NatVis files are defined in XML. There are no other representations for NatVis.

well-understood by many developers. This will help C++ developers move from
C++ to Rust. It also avoids any need to modify Rust source code. If a code
base uses code generation (such as `bindgen` or proc-macros), then standalone

> Natvis XML files would be the only way to provide visualizations for those

Reviewer: "only way" -  I'm sorry to nitpick, but using definitive language like this should probably be accompanied by justification. I would assume that with an infinite amount of time and engineers we could create at least one other way, even if it is worse?

Author: Understandable. I'll remove the last sentence since it is not essential for this RFC.

types.

### Standalone Natvis XML files

To provide standalone Natvis XML files, developers create a file with the
`.natvis` file extension. These Natvis files are then specified in the 

> `Cargo.toml` file via a new manifest key.

Reviewer: Would developers typically use tools to write these files or do it manually?

Reviewer 2: NatVis files are usually written by hand.

As an example, consider a crate with this directory structure:

```text
/Cargo.toml
  +-- src
      +-- main.rs

>       +-- main.natvis

Reviewer: Is the name important or could this be renamed to foo.natvis?

Reviewer 2: I think using the same basename as the crate's root source file has benefits, because it allows the developer to provide additional NatVis files for each crate or crate binary.

For example:

* `${crate_dir}/src/lib.natvis` - gets compiled into the primary crate for a library crate
* `${crate_dir}/src/main.natvis` - gets compiled into the primary crate for binary (exe) crate
* `${crate_dir}/src/bin/foo.natvis` - gets compiled into the `foo` binary
* `${crate_dir}/tests/big_tests.natvis` - gets compiled into the `big_tests` integration test binary
* etc.

Note that a single linker invocation can receive any number of `/NATVIS:...` parameters.

Author: As per the example above, the name of the `.natvis` file can be renamed to the developers choice as long as it has the `.natvis` extension.

Reviewer: @sivadeilra are you suggesting that Cargo should be aware of these conventions, or that this mapping can be specified by the programmer?

Author: This mapping can be specified by the developer and is not something that would be made aware to Cargo.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

This design provides a simple mechanism for cargo to collect the list of
.natvis files specified for a given crate and embed them in the resulting
pdb. It does not need any manual intervention by a Rust developer who is
consuming such a crate to get the debugging experience to work when it is
viewed under a debugger that supports the Natvis Framework.

This design does not break any existing usage of cargo or rustc. This new feature
would be strictly opt-in. The Natvis syntax may not be familiar to many Rust
developers which may lead to a period of learning the syntax. Since this feature
would be optional, a consumer of a crate that has natvis definitions for types
would not need to go through this learning curve. 

Not doing this would keep the existing debugging experience for external Rust crates.
Most Rust types, outside of the standard library, do not have any debugger views
defined for them by default which makes them difficult to interpret when viewed
under a debugger.

# Prior art
[prior-art]: #prior-art

> Many debuggers and languages already address this problem. Some do so in a way

Reviewer: I think it would be more useful to cover how other languages/compilers use natvis than how natvis itself works. IN particular how the natvis files are specified to the compiler

Reviewer: It would be useful to cover how other debuggers (GDB and LLDB in particular) do visualisation and how C/C++ compilers handle things

linker, using the `MSVC` toolchain, a `/NATVIS` linker option would be set
for each `.natvis` file. This includes `.natvis` files from all crate dependencies,
if any exist, as well as the current crate and embed them into the PDB.

The MSVC linker supports embedding debugger visualizations defined in a Natvis file
(`.natvis`) into a PDB generated by LINK through the use of the `/NATVIS` linker flag.
Since the `MSVC` linker is the only one that supports embedding Natvis files
into a PDB, this feature would be specific to the `MSVC` toolchain only.

Cargo would add new syntax in `Cargo.toml` for specifying the list of Natvis
files to be added to the crate. The new manifest key, `natvis` would be added
to a new `[debug-visualizations]` section. This would be in control of setting
the `-Z natvis` flag in rustc.

For example:

`Cargo.toml`:

```toml
cargo-features = ["natvis"]

[package]
name = "natvis"
version = "0.1.0"
edition = "2018"

[debug-visualizations]
natvis = ["a.natvis", "b.natvis"]
```

This would generate a call to rustc similar to the following,
(for simplicity purposes, most of the rustc command line has been removed):

> `rustc -Z natvis=path/to/file/a.natvis,path/to/file/b.natvis`

Reviewer 3: This should address quoting: what happens with paths containing spaces or commas?

Author: Thanks for the review. I'll update the RFC to detail quoting of paths with spaces or commas.

Author: I've added a note that paths containing spaces will need to be quoted. Since the natvis flag takes a comma-separated list of files, I've also noted that any commas within a path would be invalid in this context.

* The Windows Debugger (WinDbg)
* Visual Studio Debugger.

It should be easy for Rust developers to add debugger visualizations to their
crates.

## Supporting Natvis

This section describes how Microsoft's Natvis is supported in Rust.

To use Natvis, developers write XML documents that describe how debugger types
should be displayed using the natvis schema. (See: https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019)
The Natvis files provide patterns, which match type names, and for matching
types, a description of how to display those types. This allows for some
limited support for generic types.

Rust developers can add one or more `.natvis` files to their crate. The
`Cargo.toml` file specifies the path of these Natvis files via a new
manifest key. Cargo would then pass the set of `.natvis` files that were
specified in `Cargo.toml` and pass it to the rustc invocation for the crate
via a new `-C` flag and enabling unstable options in Rust via the
`-Z unstable-options`. See the below example for how rustc would embed these
Natvis files in the debuginfo for a binary as well as new compiler options
to be added to rustc.

> To provide Natvis files, developers create a file with the `.natvis` file

Reviewer: Since the name is specified in full in Cargo.toml, I would expect it to be unnecessary for the file to have the .natvis extension (merely convention, or for use for tools outside the Rust ecosystem)?

Reviewer 2: `.natvis` is the conventional extension used by Windows dev tools (Visual Studio, etc.) to identify NatVis files. I think it is valuable to stick with this convention.

Author: WinDbg is unable to load a Natvis file if it was embedded into the PDB via the linker without the `.natvis` extension. This is the same as how the Rust compiler currently adds Natvis for std and other libs i.e., `libstd.natvis`.

    println!("FancyRect: {:?}", fancy_rect);
}
```

and `Foo.natvis` contains:

```xml
<?xml version="1.0" encoding="utf-8"?>
<AutoVisualizer xmlns="http://schemas.microsoft.com/vstudio/debugger/natvis/2010">
    <Type Name="foo::FancyRect">
      <DisplayString>({x},{y}) + ({dx}, {dy})</DisplayString>
      <Expand>
        <Synthetic Name="LowerLeft">
          <DisplayString>({x}, {y})</DisplayString>
        </Synthetic>
        <Synthetic Name="UpperLeft">
          <DisplayString>({x}, {y + dy})</DisplayString>
        </Synthetic>
        <Synthetic Name="UpperRight">
          <DisplayString>({x + dx}, {y + dy})</DisplayString>
        </Synthetic>
        <Synthetic Name="LowerRight">
          <DisplayString>({x + dx}, {y})</DisplayString>
        </Synthetic>
      </Expand>
    </Type>
</AutoVisualizer>
```

When viewed under WinDbg, the `fancy_rect` variable would be shown as follows:

```text
> Variables:
  > fancy_rect: (10, 10) + (5, 5)
    > LowerLeft: (10, 10)
    > UpperLeft: (10, 15)
    > UpperRight: (15, 15)
    > LowerRight: (15, 10)
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> In rustc, a new `-C natvis={comma-separated list of .natvis files}` flag

Reviewer: ```suggestion
In rustc, a new `-C natvis={comma-separated list of paths}` flag
```

Reviewer 2: I would prefer to keep the `.natvis` extension here. We could clarify that the `.natvis` extension is only a convention, and that `rustc` does not require that these files have a specific extension, but it does help keep this explanatory text clear that the `-C natvis` argument value points to a specific kind of file, and not a directory.

Author: I agree, keeping this text in the documentation will clarify exactly what kind of files are expected here. 

> We could clarify that the .natvis extension is only a convention, and that rustc does not require that these files have a specific extension

@sivadeilra is there any reason we do not want to require these files to have the `.natvis` extension in rustc? WinDbg is unable to load Natvis files embedded into the PDB if they do not carry the `.natvis` extension but the linker will not emit any errors and still embed this file into the PDB.

Reviewer 2: Ah, I wasn't aware that WinDbg would not load the NatVis stream if the original file did not have a `.natvis` extension.

How about this:

1.  `rustc` doesn't care what the file extension is -- it will do what it's told, and pass through any file as a NatVis stream to the linker.
2. Cargo will only discover NatVis files that _do_ have the `.natvis` extension, in the appropriate debugger visualizations subdirectory.
3. If you use the explicit manifest key for including NatVis streams, then Cargo will check the file extension for any file explicitly specified in the manifest.  If the extension is not `.natvis`, then Cargo will print a warning, but still pass the `-C natvis` argument.

I think this will give a good development experience, but it also avoids baking in too much knowledge about what the file extension is, to rustc itself.  I'm guessing that the `.natvis` file extension requirement in WinDbg is an implementation limitation, and not an intentional choice.

Author: > I think this will give a good development experience, but it also avoids baking in too much knowledge about what the file extension is, to rustc itself.

This sounds good to me. I'll make note of the Cargo warning in the RFC.

will be added which instructs the compiler to take the set of .natvis
files for a given crate and store them in the crate metadata. This will be gated
by the existing `-Z unstable-options` flag to ensure it is only used in unstable
Rust. When running the linker, using the `MSVC` toolchain, a `/NATVIS` linker
option would be set for each `.natvis` file. This includes `.natvis` files from
all crate dependencies, if any exist, as well as the current crate and embed them
into the PDB.

The MSVC linker supports embedding debugger visualizations defined in a Natvis file
(`.natvis`) into a PDB generated by LINK through the use of the `/NATVIS` linker flag.
Since the `MSVC` linker is the only one that supports embedding Natvis files

> into a PDB, this feature would be specific to the `MSVC` toolchain only.

Reviewer: Does LLVM PDB support not include embedding natvis? If it does not and support is added in the future, I would expect that we would support natvis in the same way, right?

Author: As of this moment I do not believe embedding Natvis in a PDB is supported by LLVM. I've seen PR's in the past to add this feature but those have since been abandoned i.e., [this PR](https://reviews.llvm.org/D44328#:~:text=llvm-commits%20ruiu%20Details%20Summary%20A%20natvis%20file%20is,existing%20%22named%20stream%22%20map%20in%20the%20PDB%20file.).

Cargo would add new syntax in `Cargo.toml` for specifying the list of Natvis
files to be added to the crate. The new manifest key, `natvis` would be added

> to a new `[debug-visualizations]` section. This would be in control of setting

Reviewer: ```suggestion
to a new `[debugger-visualizations]` section. This would be in control of setting
```

Maybe? Seems like there might be confusion with the Debug trait or debug profile, etc.

Author: I'll update this to `debugger-visualizations`.

is also able to apply Natvis on the fly by loading user-specified `.natvis` files.
As with WinDbg, it also supports loading `.natvis` files that were embedded in the
PDB for a binary and automatically applying the Natvis visualizations to types from
that binary.

When using Visual Studio to build a C++ project, a developer can add a Natvis file
via the `.vcxproj` file. To add a Natvis file to a project the following can be
added to the `.vcxproj` file:

```text
<ItemGroup>
  <Natvis Include="Foo.natvis" />
</ItemGroup>
```

## GDB/LLDB

GDB and LLDB also support debugger views but in a different way than WinDbg and the
VS debugger. Natvis is not supported by either GDB or LLDB but they do support pretty
printers. Pretty printers work in the similar manner as Natvis in which they tell
the debugger to serve up a specific visualization when viewing a type in the debugger.
Pretty printers are written as python scripts and then have to be imported in to the
debugger. When a type is viewed under the debugger that has a pretty printer, that view
is automatically shown.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> Is the `[package]` section of the `Cargo.toml` manifest the best place to add this new syntax?

Reviewer: I think this question is out of date now

Author: Will remove. Thanks!

differently in the debugger.

What we _want_ is something like this:

```text
> Variables:
  > captures: {...}
    > 1: "SOME_CONSTANT"
    > 2: "42"
    > 3: "// some developer comment"
```

This RFC will describe how to support adding Natvis visualizations which is supported by:

* The Windows Debugger (WinDbg)
* Visual Studio Debugger.

It should be easy for Rust developers to add debugger visualizations to their
crates.

## Supporting Natvis

This section describes how Microsoft's Natvis is supported in Rust.

To use Natvis, developers write XML documents that describe how debugger types
should be displayed using the natvis schema. (See: https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019)
The Natvis files provide patterns, which match type names, and for matching
types, a description of how to display those types. This allows for some
limited support for generic types.

Rust developers can add one or more `.natvis` files to their crate. The
`Cargo.toml` file specifies the path of these Natvis files via a new
manifest key. Cargo would then pass the set of `.natvis` files that were

> specified in `Cargo.toml` and pass it to the rustc invocation for the crate

Author: Any crate-type that would generate a PDB should have natvis embedded. I'll add a note to the RFC to clarify this.

via a new `-C` flag and enabling unstable options in Rust via the

> `-Z unstable-options`. See the below example for how rustc would embed these

Author: The main reason this RFC requires rustc support is to ensure that any Natvis file specified for a crate would also be stored in the rmeta for that particular crate. I do not believe `link-args` are stored in the metadata today and they are also marked as an untracked option.

> `-Z unstable-options`. See the below example for how rustc would embed these

Reviewer 4: I see. This is explained in more detail just above "Drawbacks". Sorry for missing that!

I'm not sure what embedding the *paths* to the natvis files in rmeta gives us though. Since the rmetas only contain the paths, in any situation where you're distributing rlibs/rmetas you'll still need to ensure the paths point to the correct files. For example, `libstd` and `libcore` will still need special handling so that the natvis files are included with the Rust distribution artifacts. 

If the natvis files were embedded directly into rmeta, then that would resolve that issue (although I suppose we will need to extract the natvis files from the rmeta before we can pass them to `link.exe`) and the rmeta/rlib would contain everything needed by downstream dependencies. 

Alternatively, I could also imagine that Cargo would collect all the natvis files in dependencies and pass them to the final `rustc` invocation. For example, if we have a crate `binary` that depends on a crate `library`, then `cargo build` would

```bash
# compile the library dependency
rustc --crate-name library library\src\lib.rs --crate-type lib

# compile the binary, passing in all natvis files from the crate graph
rustc --crate-name binary binary\src\main.rs --crate-type bin --extern library=path/to/liblibrary.rlib --natvis library\a.natvis --natvis library\b.natvis --natvis binary\c.natvis
```

Which, if we did that, then we don't really need `--natvis` at all, we can do `-Clink-arg` as I was suggesting above.

> `-Z unstable-options`. See the below example for how rustc would embed these

Author: > If the natvis files were embedded directly into rmeta, then that would resolve that issue (although I suppose we will need to extract the natvis files from the rmeta before we can pass them to link.exe) and the rmeta/rlib would contain everything needed by downstream dependencies.

I believe that this is the correct approach here. Serializing the contents of each natvis file and embed that directly in the rmeta. This way rlibs/rmetas can be distributed and still contain the natvis for each crate. This will now add the additional step of extracting the natvis from the rmeta to ensure they can be passed into the linker.

I'll update the RFC to add this point.

    println!("FancyRect: {:?}", fancy_rect);
}
```

and `Foo.natvis` contains:

```xml
<?xml version="1.0" encoding="utf-8"?>
<AutoVisualizer xmlns="http://schemas.microsoft.com/vstudio/debugger/natvis/2010">
    <Type Name="foo::FancyRect">
      <DisplayString>({x},{y}) + ({dx}, {dy})</DisplayString>
      <Expand>
        <Synthetic Name="LowerLeft">
          <DisplayString>({x}, {y})</DisplayString>
        </Synthetic>
        <Synthetic Name="UpperLeft">
          <DisplayString>({x}, {y + dy})</DisplayString>
        </Synthetic>
        <Synthetic Name="UpperRight">
          <DisplayString>({x + dx}, {y + dy})</DisplayString>
        </Synthetic>
        <Synthetic Name="LowerRight">
          <DisplayString>({x + dx}, {y})</DisplayString>
        </Synthetic>
      </Expand>
    </Type>
</AutoVisualizer>
```

When viewed under WinDbg, the `fancy_rect` variable would be shown as follows:

```text
> Variables:
  > fancy_rect: (10, 10) + (5, 5)
    > LowerLeft: (10, 10)
    > UpperLeft: (10, 15)
    > UpperRight: (15, 15)
    > LowerRight: (15, 10)
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> In rustc, a new `-C natvis={comma-separated list of .natvis files}` flag

Author: I'll update the `-Cnatvis` argument to take a single file and allow for the flag to be specified multiple times.

a debugger. The Visual Studio Natvis framework is supported out of the box on
WinDbg. The debugger has the ability to load `.natvis` files via the `.nvload`
command and directly apply them to types within loaded modules. WinDbg is also
able to load `.natvis` files that have been embedded in the PDB for a binary and
serve up the resulting views after applying those visualizations as well. This
allows for a very smooth debugging experience which would not depend on any manual
loading of Natvis files.

## Visual Studio Debugger (VS Debugger)

The Visual Studio Debugger also supports Natvis. Similar to WinDbg, the VS Debugger
is also able to apply Natvis on the fly by loading user-specified `.natvis` files.
As with WinDbg, it also supports loading `.natvis` files that were embedded in the
PDB for a binary and automatically applying the Natvis visualizations to types from
that binary.

When using Visual Studio to build a C++ project, a developer can add a Natvis file
via the `.vcxproj` file. To add a Natvis file to a project the following can be
added to the `.vcxproj` file:

```text
<ItemGroup>
  <Natvis Include="Foo.natvis" />
</ItemGroup>
```

## GDB/LLDB

> GDB and LLDB also support debugger views but in a different way than WinDbg and the

Reviewer 4: Just to add additional context here since I saw other people asking for more information about how other debuggers work, `gdb` supports a similar concept where a special section is generated that contains the name of debugger scripts to load and execute when a binary is being debugged (see [here](https://github.com/rust-lang/rust/blob/89b9f7b284aacc5f8613438b80e4dd7bdd10549e/compiler/rustc_codegen_llvm/src/debuginfo/gdb.rs#L29-L59) where we generate this section). Gdb also seems to support embedding the scripts directly into this section as well (https://sourceware.org/gdb/onlinedocs/gdb/dotdebug_005fgdb_005fscripts-section.html).

`CrateInfo` type, `pub natvis_files: FxHashMap<StableCrateId, Vec<NatvisFile>>`.
This will allow the `MsvcLinker` type to query the list of Natvis files that exist
within the crate dependency graph and add the `/NATVIS` linker argument for each
`.natvis` file. This will also be the stage at which the Natvis files for each
crate are copied to the target directory which they will be linked from. For example,
in a debug build, the `.natvis` files will be copied to `target/debug/deps/natvis`.
Each `.natvis` file that is copied over to the new directory will have a new name to
ensure it is unique across `.natvis` files for all crates. The naming scheme will be
`<crate_name>-<stable_crate_id>-<file_name>-<count>.natvis`.

# Drawbacks
[drawbacks]: #drawbacks

One drawback here is that a lot of types implement the Debug trait which
already specifies how a type should be viewed when debugging. Implementing
this RFC would mean a Rust developer would have to manually specify the
Natvis for a type that may already have implemented the Debug trait which
would be redundant. Currently, running the Debug trait in the debugger directly

> is not possible and so a manual definition would be required to have a debugger view.

Reviewer 5: I'm glad you mentioned this downside, because when I read the idea at the start of this RFC, it was the first thing that came to my mind. We could trivially have miri execute the MIR of a Debug impl. The only unimplemented part of this is asking the debugger for the parts of the real memory needed to run the Debug impl.

Instead of opening the road to supporting N (there will be more) debug visualization schemes (when we already derive or implement one), we could totally go down the road of adding the infrastructure for invoking miri. I'd be happy to iterate on this idea and help writing an RFC for it. I think it would be much more useful, as it would not require people to actually change their crates, we get this feature in hindsight to existing crates.

Reviewer 5: When doing this via miri or some derivative of it we can actually be more powerful than plain visualization. We can inform about nested debug formatting runs (allowing a debugger to go from a nested visualization to its memory) and generally use the `fmt::Formatter` API to create structured output instead of just a stringification.

This would generate a call to rustc similar to the following,
(for simplicity purposes, most of the rustc command line has been removed):

`rustc -C natvis=path/to/file/a.natvis -C natvis=path/to/file/b.natvis -Z unstable-options`

`.natvis` files that contain spaces or commas within the path will be quoted to ensure
the path remains valid when passing command line options to rustc.

The `CrateRoot` type would also need to be updated to account for `.natvis`
files for crates within the dependency graph. To reflect this a new type
`pub struct NatvisFile` will be created to ensure the contents of a `.natvis`
file can be serialized and stored in the crate metadata. The `CrateRoot` would

> contian the field, `natvis_files: Lazy<[NatvisFile]>`.

Reviewer 4: nit: `contain`

Another change that would need to be made here is to add a new field to the

> `CrateInfo` type, `pub used_crates_natvis: Option<Vec<PathBuf>>`. This will

Reviewer 4: I don't think this is needed anymore? We can just use `natvis_files` directly.

Author: We still need this to be able to add the `/NATVIS` linker argument in the debug info. At this stage we do not have access to the `CrateMetadata` or the `TyCtx`. The `CrateInfo` is accessible which is why this field would still be needed.

would be optional, a consumer of a crate that has natvis definitions for types
would not need to go through this learning curve. 

## Alternatives

### Alternative 1: existing -C link-arg flag

Supporting this option would mean that changes to rustc are not necessary.
The changes would be limited to Cargo, which would be responsible for collecting
the set of Natvis files and passing `-Clink-arg=/NATVIS:{file-path}` for each
Natvis file.

The drawbacks for this option is that it will only collect Natvis files for the
top-most manifest. This will not walk the dependency graph and find all relevant
Natvis files so this will only work for targets that produce a `DLL` or `EXE` and
not an `.rlib`.

### Alternative 2: custom build script to set /NATVIS linker flag

Supporting this option would mean that changes to cargo and rustc are not necessary.
Each individual crate would be able to create a custom build script that would set
the rustc `link-arg` flag `cargo:rustc-link-arg=/NATVIS:{file-path}` for each Natvis
file.

> The drawbacks for this option is that it would force all Rust developers to manually

Reviewer 5: Well, those that have to add natvis files also need to do some work, the build script seems like the smallest work here. A build script is still annoying, I fully agree with that

Author: > the build script seems like the smallest work here. A build script is still annoying, I fully agree with that.

This is true but as with my above comment, I'm not sure the build script would be able to account for transitive dependencies. Using the `link-arg` flag would work if the current crate being built is not an `rlib` but once again that would only add Natvis solely for the current crate.

> create a build script and ensure it is kept up-to-date whenever the set of Natvis files

Reviewer 5: A build script could auto detect the natvis files. Paired with a crate that does all the heavy lifting and just offers an `import_natvis` function this seems like a small drawback

Reviewer 5: Yea, that's true. Please put it into the doc so the problem about transitive dependencies doesn't get lost

are updated. This also would set the linker argument regardless of the linker being used.

> This could lead to potential build errors if the linker being used returns an error for

Reviewer 5: Hmm... don't we already have this situation for all uses of linker flags? I am certain I have seen detection logic to only set msvc flags on the right target

Author: This I am not sure about as I do not have much experience around build scripts. I will look into this to see how this can be done.

Reviewer 5: Well, the point is moot due to the other concerns you raised, so maybe just remove the comment about linker flags on the wrong platform?

`.natvis` file, there may become an issue with the visibility of the source code.
Also, if/when other debugger visualization formats are supported, it could become
very obscure to read the source with large amounts of visualization scripts from
multiple schemas all being directly embedded in source code.

### Alternative 4: miri executes the MIR of a Debug impl within a debugger

Supporting this option would mean that changes to cargo and rustc are not necessary.
This would have the added benefit of taking full advantage of existing implementations
of the `Debug` trait. Many Rust developers already implement the `Debug` trait which is
used to format how types should be viewed, this would only ease the debugging quality of
Rust when viewed under any debugger. This option also has the added benefit of not
requiring any changes to a crate from a Rust developer by leveraging existing `Debug` impls.

The drawbacks for this option is that this has not been fully investigated to
determine its viability. This could be a great potential feature to ease
debugging Rust but without concrete data to push this towards a potential RFC,

> I would assume supporting debugging in the systems that are already heavily used

Reviewer 5: I agree, but i am worried we'll settle for "ok instead of great" and never put any work into this without the pressure of improving debug visualizations

Author: I think this is always something of a concern when planning features. The feature this RFC is trying to implement is pretty well-defined and has clear value. The miri approach has yet to be investigated and while it may a great long-term solution, I think it needs to be investigated to ensure it can solve the problems mentioned above. I don't know if we want to block this feature on such an investigation into the miri solution.

Reviewer 5: Yea... not sure who has the time to put in the work for that XD which is kind of my point. If we block it in that investigation, everyone who wants natvis has motivation to do the investigation.

But... I would probably feel annoyed if someone did that do me, so I wouldn't feel good to use that as leverage.
